website/docs/proposals/20250620-auto-scaler-for-inference-workloads.md

```diff
@@ -0,0 +1,262 @@
+
+To ensure ease of use, The specialized auto-scaler is hosted in a independent repo(kaito-project/llm-auto-scaler). at the same time, the llm-auto-scaler component can work with kaito without depending on any third-party components. 
+
+### Goals
+
+- llm-auto-scaler is a specialized auto-scaler for scaling gpu workloads automatically, and can integrate with kaito to work.
+- It is flexible to support mulitple scale strategies, and only one basic scale strategy(scaling workloads according to metrics change) is supported in the first version.
+
+### Non-Goals/Future Work
+
+- Support cron scale strategy(like cron job) for llm-auto-sacler in future version.
+- Only support to configure one metric for basic scale strategy, mutiple metrics will be supported in future version.
+- scale subresource api for workspace CRD is not covered in this proposal.
+- The time efficiency of the auto-scaler is not within the scope of this proposal, as it is influenced by mutliple external factors, including GPU node provisioning, LLM image pulling, etc.
+
+## Proposal
+
+### Auto-Scaler Architecture
+
+The llm-auto-scaler component scrapes metrics from inference pod according to configurations in LLMAutoScaler CRD, and scaler controller calculate desired replicas by integrating scraped metrics and scale strategy,
+then scale workspace replicas through /scale subresource API. The detailed auto-scaler architecture is shown in the following figure:
+
+![auto-scaler](../../static/img/llm-auto-scaler.png)
+
+- LLMAutoScaler CRD: is used as auto-scaler configuration(including scale strategy, target reference, etc.) for specified workspace resource.
+- Metrics Scraper: a module in auto-scaler of llm-auto-scaler and used for scraping metrics from inference pod.
+- Scale Strategy: is used to specify scaler, like basic scaler, cron scaler. different scaler have different algorithm to calculate desired replicas. In this proposal, only basic scaler will be supported. and more strategies will be supported in future versions.
+- Scaler Controller: is used for integrating metrics scraper and scaler strategy, also including invoke scale subresource API of workspace.
+
+### LLMAutoScaler CRD
+
+```
+type ProtocolType string
+
+const (
+	HTTP  ProtocolType = "http"
+	HTTPS ProtocolType = "https"
+)
+
+// MetricIdentifier defines the way to fetch the specific metric
+type MetricIdentifier struct {
+	// Name identifies the specific metric to monitor.
+	// If unset, vllm:num_requests_waiting will be used.
+	Name string
+
+	// selector is the string-encoded form of a standard kubernetes label selector for the given metric.
+	// if unset, a selector related to ScaleTargetRef will be configured(like workspace specified labels).
+	Selector *metav1.LabelSelector
```

Comment:
I don't think we need this if we implement [labelSelectorPath](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource) from scale subresource in KAITO's Workspace

Reply:
I'wll remove it.